File: archive/python-legacy/polymarket/data_pipeline/openweathermap_fetcher.py
"""
OpenWeatherMap data fetcher module
Retrieves historical weather data from OpenWeatherMap API
Requires API key, global coverage
"""

import logging
import requests
import pandas as pd
from datetime import datetime, timedelta
from typing import List, Dict, Optional

from polymarket.config import OPENWEATHERMAP_API_KEY

logger = logging.getLogger(__name__)

LOCATIONS = {
    "NYC": {"lat": 40.71, "lon": -74.01, "name": "New York, NY"},
    "LA": {"lat": 34.05, "lon": -118.24, "name": "Los Angeles, CA"},
    "London": {"lat": 51.51, "lon": -0.13, "name": "London, UK"},
    "Chicago": {"lat": 41.88, "lon": -87.63, "name": "Chicago, IL"},
    "Dallas": {"lat": 32.78, "lon": -96.80, "name": "Dallas, TX"},
    "Denver": {"lat": 39.74, "lon": -104.99, "name": "Denver, CO"},
    "Miami": {"lat": 25.76, "lon": -80.19, "name": "Miami, FL"},
    "Boston": {"lat": 42.36, "lon": -71.06, "name": "Boston, MA"},
}

BASE_URL = "https://api.openweathermap.org/data/3.0/onecall/day_summary"


class OpenWeatherMapFetcher:
    """Fetches historical weather data from OpenWeatherMap API"""

    def __init__(self, api_key: str = OPENWEATHERMAP_API_KEY):
        self.api_key = api_key
        self.session = requests.Session()

    def is_available(self) -> bool:
        """Check if API key is configured"""
        return bool(self.api_key)

    def fetch_daily_observations(
        self,
        latitude: float,
        longitude: float,
        start_date: str,
        end_date: str,
        location_id: str = "OWM",
    ) -> pd.DataFrame:
        """
        Fetch daily weather observations for a location

        Args:
            latitude: Location latitude
            longitude: Location longitude
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)
            location_id: Identifier for the location

        Returns:
            DataFrame with standardized weather observations
        """
        if not self.is_available():
            logger.info("OpenWeatherMap: No API key configured, skipping")
            return pd.DataFrame()

        start = datetime.strptime(start_date, "%Y-%m-%d")
        end = datetime.strptime(end_date, "%Y-%m-%d")
        observations = []

        current = start
        while current <= end:
            date_str = current.strftime("%Y-%m-%d")
            params = {
                "lat": latitude,
                "lon": longitude,
                "date": date_str,
                "appid": self.api_key,
                "units": "metric",
            }

            try:
                response = self.session.get(BASE_URL, params=params, timeout=15)
                response.raise_for_status()
                data = response.json()

                temp = data.get("temperature", {})
                precip = data.get("precipitation", {})
                wind = data.get("wind", {})

                observations.append({
                    "date": pd.Timestamp(date_str),
                    "station_id": f"OWM_{location_id}",
                    "temperature_max": temp.get("max"),
                    "temperature_min": temp.get("min"),
                    "temperature_mean": (
                        (temp.get("max", 0) + temp.get("min", 0)) / 2
                        if temp.get("max") is not None and temp.get("min") is not None
                        else None
                    ),
                    "precipitation_total": precip.get("total", 0),
                    "wind_speed_mean": wind.get("max", {}).get("speed"),
                })

            except requests.RequestException as e:
                logger.debug(f"OpenWeatherMap error for {date_str}: {e}")

            current += timedelta(days=1)

        if not observations:
            return pd.DataFrame()

        df = pd.DataFrame(observations)
        logger.info(f"OpenWeatherMap: Fetched {len(df)} observations for {location_id}")
        return df

    def fetch_location(
        self,
        location_key: str,
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for a pre-configured location"""
        location = LOCATIONS.get(location_key)
        if not location:
            logger.error(f"Unknown location: {location_key}")
            return pd.DataFrame()

        return self.fetch_daily_observations(
            latitude=location["lat"],
            longitude=location["lon"],
            start_date=start_date,
            end_date=end_date,
            location_id=location_key,
        )

    def fetch_multiple_locations(
        self,
        location_keys: List[str],
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for multiple locations and combine"""
        all_data = []
        for key in location_keys:
            df = self.fetch_location(key, start_date, end_date)
            if not df.empty:
                all_data.append(df)

        if all_data:
            return pd.concat(all_data, ignore_index=True)
        return pd.DataFrame()

File: src/data_pipeline/openweathermap_fetcher.rs
use std::collections::HashMap;

use chrono::{Duration, NaiveDate};
use serde_json::Value;

use crate::config::openweathermap_api_key;
use crate::types::WeatherRecord;

const BASE_URL: &str = "https://api.openweathermap.org/data/3.0/onecall/day_summary";

#[derive(Clone)]
pub struct OpenWeatherMapFetcher {
    api_key: String,
    client: reqwest::blocking::Client,
    locations: HashMap<String, (f64, f64)>,
}

impl OpenWeatherMapFetcher {
    pub fn new(api_key: Option<String>) -> Self {
        Self {
            api_key: api_key.unwrap_or_else(openweathermap_api_key),
            client: crate::data_pipeline::build_client(15),
            locations: default_locations(),
        }
    }

    pub fn is_available(&self) -> bool {
        !self.api_key.is_empty()
    }

    pub fn fetch_daily_observations(
        &self,
        latitude: f64,
        longitude: f64,
        start_date: NaiveDate,
        end_date: NaiveDate,
        location_id: &str,
    ) -> Vec<WeatherRecord> {
        if !self.is_available() {
            return Vec::new();
        }

        let mut out = Vec::new();
        let mut current = start_date;

        while current <= end_date {
            let query = [
                ("lat", latitude.to_string()),
                ("lon", longitude.to_string()),
                ("date", current.to_string()),
                ("appid", self.api_key.clone()),
                ("units", "metric".to_string()),
            ];

            if let Ok(resp) = self.client.get(BASE_URL).query(&query).send() {
                if let Ok(json) = resp.json::<Value>() {
                    let mut row = WeatherRecord::new(current, format!("OWM_{location_id}"));
                    row.temperature_max = json
                        .get("temperature")
                        .and_then(|v| v.get("max"))
                        .and_then(|v| v.as_f64());
                    row.temperature_min = json
                        .get("temperature")
                        .and_then(|v| v.get("min"))
                        .and_then(|v| v.as_f64());

                    row.temperature_mean = match (row.temperature_max, row.temperature_min) {
                        (Some(mx), Some(mn)) => Some((mx + mn) / 2.0),
                        _ => None,
                    };

                    row.precipitation_total = json
                        .get("precipitation")
                        .and_then(|v| v.get("total"))
                        .and_then(|v| v.as_f64())
                        .or(Some(0.0));
                    row.wind_speed_mean = json
                        .get("wind")
                        .and_then(|v| v.get("max"))
                        .and_then(|v| v.get("speed"))
                        .and_then(|v| v.as_f64());
                    out.push(row);
                }
            }

            current += Duration::days(1);
        }

        out
    }

    pub fn fetch_location(&self, location_key: &str, start_date: NaiveDate, end_date: NaiveDate) -> Vec<WeatherRecord> {
        let Some((lat, lon)) = self.locations.get(location_key) else {
            return Vec::new();
        };
        self.fetch_daily_observations(*lat, *lon, start_date, end_date, location_key)
    }
}

impl Default for OpenWeatherMapFetcher {
    fn default() -> Self {
        Self::new(None)
    }
}

fn default_locations() -> HashMap<String, (f64, f64)> {
    [
        ("NYC", (40.71, -74.01)),
        ("LA", (34.05, -118.24)),
        ("London", (51.51, -0.13)),
        ("Chicago", (41.88, -87.63)),
        ("Dallas", (32.78, -96.80)),
        ("Denver", (39.74, -104.99)),
        ("Miami", (25.76, -80.19)),
        ("Boston", (42.36, -71.06)),
    ]
    .into_iter()
    .map(|(k, v)| (k.to_string(), v))
    .collect()
}
